src/ott/neural/models/base_solver.py

```diff
@@ -0,0 +1,485 @@
+        y_quad: jnp.ndarray
+    ) -> Tuple[jnp.ndarray, jnp.ndarray, jnp.ndarray, jnp.ndarray]:
+      geom = pointcloud.PointCloud(
+          x_lin, y_lin, epsilon=epsilon, scale_cost=scale_cost, cost_fn=cost_fn
+      )
+      return ot_solver(
+          linear_problem.LinearProblem(geom, tau_a=tau_a, tau_b=tau_b)
+      ).matrix
+
+    return match_pairs_filtered if filter_input else match_pairs
+
+  def _get_gromov_match_fn(
+      self,
+      ot_solver: Any,
+      cost_fn: Union[Any, Mapping[str, Any]],
+      scale_cost: Union[Union[bool, int, float,
+                              Literal["mean", "max_norm", "max_bound",
+                                      "max_cost", "median"]],
+                        Dict[str, Union[bool, int, float,
+                                        Literal["mean", "max_norm", "max_bound",
+                                                "max_cost", "median"]]]],
+      tau_a: float,
+      tau_b: float,
+      fused_penalty: float,
+  ) -> Callable:
+    if isinstance(cost_fn, Mapping):
+      assert "cost_fn_xx" in cost_fn
+      assert "cost_fn_yy" in cost_fn
+      cost_fn_xx = cost_fn["cost_fn_xx"]
+      cost_fn_yy = cost_fn["cost_fn_yy"]
+      if fused_penalty > 0:
+        assert "cost_fn_xy" in cost_fn_xx
+        cost_fn_xy = cost_fn["cost_fn_xy"]
+    else:
+      cost_fn_xx = cost_fn_yy = cost_fn_xy = cost_fn
+
+    if isinstance(scale_cost, Mapping):
+      assert "scale_cost_xx" in scale_cost
+      assert "scale_cost_yy" in scale_cost
+      scale_cost_xx = scale_cost["scale_cost_xx"]
+      scale_cost_yy = scale_cost["scale_cost_yy"]
+      if fused_penalty > 0:
+        assert "scale_cost_xy" in scale_cost
+        scale_cost_xy = cost_fn["scale_cost_xy"]
+    else:
+      scale_cost_xx = scale_cost_yy = scale_cost_xy = scale_cost
+
+    @jax.jit
+    def match_pairs(
+        x_lin: Optional[jnp.ndarray],
+        x_quad: Tuple[jnp.ndarray, jnp.ndarray],
+        y_lin: Optional[jnp.ndarray],
+        y_quad: Tuple[jnp.ndarray, jnp.ndarray],
+    ) -> Tuple[jnp.array, jnp.array]:
+      geom_xx = pointcloud.PointCloud(
+          x=x_quad, y=x_quad, cost_fn=cost_fn_xx, scale_cost=scale_cost_xx
+      )
+      geom_yy = pointcloud.PointCloud(
+          x=y_quad, y=y_quad, cost_fn=cost_fn_yy, scale_cost=scale_cost_yy
+      )
+      if fused_penalty > 0:
+        geom_xy = pointcloud.PointCloud(
+            x=x_lin, y=y_lin, cost_fn=cost_fn_xy, scale_cost=scale_cost_xy
+        )
+      else:
+        geom_xy = None
+      prob = quadratic_problem.QuadraticProblem(
+          geom_xx,
+          geom_yy,
+          geom_xy,
+          fused_penalty=fused_penalty,
+          tau_a=tau_a,
+          tau_b=tau_b
+      )
+      out = ot_solver(prob)
+      return out.matrix
+
+    return match_pairs
+
+
+class UnbalancednessMixin:
+  """Mixin class to incorporate unbalancedness into neural OT models."""
+
+  def __init__(
+      self,
+      rng: jax.Array,
```

Comment:
In general, I don't think random state should be part of this.

Reply:
How would you handle it otherwise? Would you pass it to the function `resample..`?